From 7f8bf7ae9941bfbb976920e4f1f63831feb1a8c9 Mon Sep 17 00:00:00 2001
From: Girija P <gpandiya@qti.qualcomm.com>
Date: Thu, 8 Jan 2026 00:17:05 +0530
Subject: [PATCH 19/54] move prplmesh persistency setting logic after parsing
 all bss_config Create launcher for prplmesh-agent to remove mld entrant file
 Workaround for deriving unique bssid for new links/vaps in MLDcase

Signed-off-by: Girija P <gpandiya@qti.qualcomm.com>
---
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       | 54 ++++++++++---------
 common/beerocks/scripts/CMakeLists.txt        |  8 +++
 .../scripts/beerocks_agent_launcher.sh.in     |  9 ++++
 3 files changed, 45 insertions(+), 26 deletions(-)
 create mode 100755 common/beerocks/scripts/beerocks_agent_launcher.sh.in

diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index 6b02d19be..d037a8ab9 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -1464,29 +1464,6 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
                     conf.remove_bss_config(vap_id);
                 }
 
-                if (new_radio_iface_name_update_req != 0 && !new_radio_iface_name.empty()) {
-                    // Extract the suffix from the original iface name
-
-                    //update prplmesh config file for persistance
-                    const std::string package_name = "prplmesh";
-                    const std::string section_type = "wifi-device";
-                    const std::string option_name  = "hostap_iface";
-
-                    beerocks::bpl::cfg_update_prplmesh_option(package_name, section_type, option_name, base_radio_iface_name, new_radio_iface_name);
-
-                    std::string link_suffix;
-                    size_t link_pos = m_radio_info.iface_name.find("_link");
-                    if (link_pos != std::string::npos) {
-                        link_suffix = m_radio_info.iface_name.substr(link_pos);
-                    }
-
-                    // Append the suffix to the new base name
-                    std::string updated_iface_name = new_radio_iface_name + link_suffix;
-
-                    LOG(INFO) << "Updating m_radio_info.iface_name to " << updated_iface_name;
-                    m_radio_info.iface_name = updated_iface_name;
-                }
-
 
                 if (conf.update_required()) {
                     if (!conf.store()) {
@@ -1512,6 +1489,29 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
 
     }
 
+    if (new_radio_iface_name_update_req != 0 && !new_radio_iface_name.empty()) {
+	    // Extract the suffix from the original iface name
+
+	    //update prplmesh config file for persistance
+	    const std::string package_name = "prplmesh";
+	    const std::string section_type = "wifi-device";
+	    const std::string option_name  = "hostap_iface";
+
+	    beerocks::bpl::cfg_update_prplmesh_option(package_name, section_type, option_name, base_radio_iface_name, new_radio_iface_name);
+
+	    std::string link_suffix;
+	    size_t link_pos = m_radio_info.iface_name.find("_link");
+	    if (link_pos != std::string::npos) {
+		    link_suffix = m_radio_info.iface_name.substr(link_pos);
+	    }
+
+	    // Append the suffix to the new base name
+	    std::string updated_iface_name = new_radio_iface_name + link_suffix;
+
+	    LOG(INFO) << "Updating m_radio_info.iface_name to " << updated_iface_name;
+	    m_radio_info.iface_name = updated_iface_name;
+    }
+
     if (abort) {
         LOG(ERROR) << "Autoconfiguration:failed: ";
         return false;
@@ -2267,6 +2267,7 @@ int ap_wlan_hal_nl80211::add_bss(std::string &ifname, son::wireless_utils::sBssI
                << "BSS configuration: " << conf;
 
     //cmd << "ADD bss_config=" << get_iface_name() << ":" << conf_path;
+
     cmd << "ADD bss_config=" <<m_radio_info.phy_name<< ":" <<conf_path;
     if (!wpa_ctrl_send_msg(cmd.str(), global_iface)) {
         LOG(ERROR) << "Failed to create the new BSS!";
@@ -2509,10 +2510,11 @@ void ap_wlan_hal_nl80211::generate_new_bssid(uint16_t new_vap_index,sMacAddr &ne
         mbssid_group_size = 4;
     }
 
-    if (m_radio_info.sta_enabled) {
+//    if (m_radio_info.sta_enabled) {
         new_vap_index += 1;
-        LOG(INFO) << "STA intreface is present in " << m_radio_info.iface_name << " new_vap index: " << new_vap_index;
-    }
+	LOG(INFO) << "Workaround: Incrementing bssid index +2 irrespective of sta status";
+//        LOG(INFO) << "STA intreface is present in " << m_radio_info.iface_name << " new_vap index: " << new_vap_index;
+//    }
 
     if (wpa_ctrl_send_msg("STATUS", reply)) {
         LOG(INFO) << __func__ << " STATUS Success";
diff --git a/common/beerocks/scripts/CMakeLists.txt b/common/beerocks/scripts/CMakeLists.txt
index f19114eca..eec8e8190 100644
--- a/common/beerocks/scripts/CMakeLists.txt
+++ b/common/beerocks/scripts/CMakeLists.txt
@@ -11,11 +11,19 @@ configure_file(
         @ONLY
         )
 
+configure_file(
+        "${CMAKE_CURRENT_SOURCE_DIR}/beerocks_agent_launcher.sh.in"
+        "${CMAKE_CURRENT_BINARY_DIR}/beerocks_agent_launcher.sh"
+        @ONLY
+        )
+
 set(scripts ${CMAKE_CURRENT_BINARY_DIR}/prplmesh_utils.sh)
 
 if (NOT TARGET_PLATFORM STREQUAL "linux")
 list(APPEND scripts ${CMAKE_CURRENT_BINARY_DIR}/beerocks_watchdog.sh)
 endif()
 
+list(APPEND scripts ${CMAKE_CURRENT_BINARY_DIR}/beerocks_agent_launcher.sh)
+
 install(PROGRAMS ${scripts} DESTINATION scripts)
 file(COPY ${scripts} DESTINATION ${CMAKE_MULTIAP_OUTPUT_DIRECTORY}/scripts)
diff --git a/common/beerocks/scripts/beerocks_agent_launcher.sh.in b/common/beerocks/scripts/beerocks_agent_launcher.sh.in
new file mode 100755
index 000000000..bf3053ce8
--- /dev/null
+++ b/common/beerocks/scripts/beerocks_agent_launcher.sh.in
@@ -0,0 +1,9 @@
+#!/bin/sh
+#Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+#SPDX-License-Identifier: ISC
+
+PRPLMESH_DIR="/opt/prplmesh/"
+
+rm -f /tmp/beerocks/mld_pair_entrant.txt
+$PRPLMESH_DIR/bin/beerocks_agent
+
-- 
2.34.1

